### PR TITLE
PW-2524 - Update 3DS1 Order Flow

### PR DESCRIPTION
**Pull request by martinsrenato**
- Creating ON_HOLD order when REDIRECT_SHOPPER response is received
- Update payment info and order status when payment is completed or cancelled

### CODE REVIEW

File: adyenv6core/src/com/adyen/v6/facades/DefaultAdyenCheckoutFacade.java [hunk 1]
```diff
@@ -712,9 +749,20 @@ private OrderData createOrderFromPaymentsResponse(final PaymentsResponse payment
...
         }
+
         return orderData;
     }
 
+    private OrderData placePendingOrder(PaymentsResponse.ResultCodeEnum resultCode) throws InvalidCartException {
+        OrderData orderData = getCheckoutFacade().placeOrder();
+
+        OrderModel orderModel = orderRepository.getOrderModel(orderData.getCode());
+        orderModel.setStatus(OrderStatus.ON_HOLD);
```
Comment (rkewlani) [inline adyenv6core/src/com/adyen/v6/facades/DefaultAdyenCheckoutFacade.java:760]: OrderStatus ON_HOLD has a different meaning in order flow. On Hold means -Sourcing Failed. We may need to introduce a different status.

Comment (martinsrenato) [reply adyenv6core/src/com/adyen/v6/facades/DefaultAdyenCheckoutFacade.java]: Updated to PAYMENT_PENDING

File: adyenv6core/src/com/adyen/v6/facades/AdyenCheckoutFacade.java [hunk 2]
```diff
@@ -223,4 +223,5 @@
      * Check POS Payment status using Adyen Terminal API
      */
     OrderData checkPosPaymentStatus(HttpServletRequest request, CartData cartData) throws Exception;
+
```
Comment (rkewlani) [inline adyenv6core/src/com/adyen/v6/facades/AdyenCheckoutFacade.java:226]: is it intentional ?

File: adyenv6core/src/com/adyen/v6/facades/DefaultAdyenCheckoutFacade.java [hunk 3]
```diff
@@ -550,11 +568,26 @@ public OrderData handle3DResponse(final HttpServletRequest request) throws Excep
...
             }
 
-            restoreSessionCart();
             PaymentsResponse paymentsResponse = getAdyenPaymentService().authorise3DPayment(sessionPaymentData, paRes, md);
+
+            String orderCode = paymentsResponse.getMerchantReference();
+            OrderModel orderModel = retrieveOrder(orderCode);
+
+            updateOrderPaymentInfo(paymentsResponse, orderModel);
+
```
Comment (rkewlani) [inline adyenv6core/src/com/adyen/v6/facades/DefaultAdyenCheckoutFacade.java:577]: We can move this to later stage where we are setting order status to make sure we are updating all order related details at one place.

File: adyenv6core/src/com/adyen/v6/facades/DefaultAdyenCheckoutFacade.java [hunk 4]
```diff
@@ -1235,6 +1283,47 @@ private boolean isPosTimedOut(HttpServletRequest request) {
...
+        getSessionService().removeAttribute(PAYMENT_METHOD);
+
+        return orderModel;
+    }
+
+    private void restoreCartFromOrder(String orderCode) {
+        OrderModel orderModel = getOrderRepository().getOrderModel(orderCode);
+        if (orderModel == null) {
+            LOGGER.error("Could not restore cart to session, order not found!");
+            getCartService().removeSessionCart();
```
Comment (rkewlani) [inline adyenv6core/src/com/adyen/v6/facades/DefaultAdyenCheckoutFacade.java:1306]: which cart will be present in session during this execution is also important.

File: adyenv6core/testsrc/com/adyen/v6/facade/AdyenCheckoutFacadeTest.java [hunk 5]
```diff
@@ -376,4 +411,135 @@ private void setSuccessfulPaymentStubs() throws InvalidCartException {
         //updateOrder
         doNothing().when(adyenOrderService).updateOrderFromPaymentsResponse(any(), any());
     }
+
```
Comment (rkewlani) [inline adyenv6core/testsrc/com/adyen/v6/facade/AdyenCheckoutFacadeTest.java:414]: can we add more test cases specifically for new order status PENDING_PAYMENT

File: adyenv6core/src/com/adyen/v6/facades/AdyenCheckoutFacade.java (hunk 2 above)
Comment (martinsrenato) [reply adyenv6core/src/com/adyen/v6/facades/AdyenCheckoutFacade.java:226]: no, I'll revert

File: adyenv6core/src/com/adyen/v6/facades/DefaultAdyenCheckoutFacade.java (hunk 4 above)
Comment (martinsrenato) [reply adyenv6core/src/com/adyen/v6/facades/DefaultAdyenCheckoutFacade.java:1306]: Not sure if I understand your point, 
but I think I can remove the `getCartService().removeSessionCart();` and the current cart (if any) will remain in session... is that what you mean?

File: adyenv6core/src/com/adyen/v6/facades/DefaultAdyenCheckoutFacade.java (hunk 3 above)
Comment (martinsrenato) [reply adyenv6core/src/com/adyen/v6/facades/DefaultAdyenCheckoutFacade.java:577]: Makes sense, will do

File: adyenv6core/testsrc/com/adyen/v6/facade/AdyenCheckoutFacadeTest.java (hunk 5 above)
Comment (martinsrenato) [reply adyenv6core/testsrc/com/adyen/v6/facade/AdyenCheckoutFacadeTest.java:415]: `testAuthorisePaymentRedirect` already goes through that flow. I added some more assertions to make it clear.

File: adyenv6core/src/com/adyen/v6/facades/DefaultAdyenCheckoutFacade.java (hunk 4 above)
Comment (martinsrenato) [reply adyenv6core/src/com/adyen/v6/facades/DefaultAdyenCheckoutFacade.java:1374]: Fixed that on branch PW-2530-restore-cart

File: adyenv6core/src/com/adyen/v6/facades/DefaultAdyenCheckoutFacade.java [hunk 6]
```diff
@@ -1316,6 +1351,47 @@ public OrderData handleComponentResult(String resultJson) throws Exception {
...
+            //TODO change exception
+            throw new InvalidCartException("Order does not exist!");
+        }
+
+        getSessionService().removeAttribute(SESSION_LOCKED_CART);
+        getSessionService().removeAttribute(SESSION_PAYMENT_DATA);
+        getSessionService().removeAttribute(THREEDS2_FINGERPRINT_TOKEN);
+        getSessionService().removeAttribute(THREEDS2_CHALLENGE_TOKEN);
+        getSessionService().removeAttribute(PAYMENT_METHOD);
+
```
Comment (Aleffio) [inline adyenv6core/src/com/adyen/v6/facades/DefaultAdyenCheckoutFacade.java:1366]: do we need to remove `SESSION_MD` as well or is it not necessary/applicable?

Comment (martinsrenato) [reply adyenv6core/src/com/adyen/v6/facades/DefaultAdyenCheckoutFacade.java:1366]: makes sense, will add that

Comment (martinsrenato) [reply adyenv6core/src/com/adyen/v6/facades/DefaultAdyenCheckoutFacade.java:1366]: we're working on this method on a separate branch
pushed there https://github.com/Adyen/adyen-hybris/tree/PW-2530-restore-cart

**Pull request merged**